### DISCUSSION TITLE
Set function for Record type value

### DISCUSSION THREAD

**Discussion by cchannon** [microsoft/Power-Fx#889] [Q&A] [answered]
In the testing I've done so far, I can't seem to update Record type objects with mutation functions. Maybe there's some other flag I need to set to make this possible?

So here's the idea: I have a Record, and it has three objects in it (all strings, to keep things simple):

`{ Id: "My Id", Name: "My Name", LogicalName: "MyLogicalName" }`

I declare a var for it using FromJson:

```
engine.UpdateVariable("myobj", FormulaValue.FromJson(
                JsonSerializer.Serialize(
                    new EntityRefObj(
                        "this is an id",
                        "this is a name",
                        "this is a logical name"
                    )
                )
            ));
```

and the record and its member objects are all accessible to me. BUT, when I go to set one:

`Set(myobj, Patch(myobj, {Id: "A New and Different Id"}))"`

I get the error "Invalid argument type." referring to the Patch method. The Patch on its own outputs exactly what I'd expect (a new Record Object that is a match for the old one but has an updated Id) but Set seems unable to accept its output.

I thought maybe it was just the type of Var I was declaring (because of the FromJson) so I also tried it with NewRecordFromFields, but the results are the same

```
NamedValue[] values = new NamedValue[] {
                new NamedValue(new KeyValuePair<string, FormulaValue>("Id", FormulaValue.New("this is an id"))),
                new NamedValue(new KeyValuePair<string, FormulaValue>("Name", FormulaValue.New("this is a Name"))),
                new NamedValue(new KeyValuePair<string, FormulaValue>("LogicalName", FormulaValue.New("this is a LogicalName")))
            };

            engine.UpdateVariable("thingy", FormulaValue.NewRecordFromFields(values));
```

So what's the piece I am missing here? How am I supposed to mutate a Record type Var, if not as documented [here](https://learn.microsoft.com/en-us/power-platform/power-fx/reference/function-set#examples)?

**Comment by cchannon** [comment]
Here, I think this screengrab is a really good illustration of the problem I am seeing:

![image](https://user-images.githubusercontent.com/12833126/206283923-d665f84d-20fe-4203-93ba-56075dda889a.png)

I can Set() a Record and feed it a new record manually, no problem. 
I can Patch() a Record and get back a new record, no problem.
But I can't pass the output of Patch() to Set() in order to be able to update a Record without having to redeclare all its members

**Comment by jorisdg** [comment]
What version of the codebase are you using? There was a PR 16 days ago that changed the return type of Patch to Record, as it was returning Table prior to that (which would be an invalid argument type as you're seeing).

https://github.com/microsoft/Power-Fx/pull/853

**Reply by cchannon** [reply]
I am on 0.2.3 for both Core and Interpreter, which is the latest I can get from Nuget. Should I just be pulling straight from the repo instead?

![image](https://user-images.githubusercontent.com/12833126/206319093-3c355151-73ab-4df4-b4e9-6334a6fc53dc.png)

**Reply by MikeStall** [reply]
@cchannon - I'd recommend pulling the daily builds from here. This will be a nuget _per commit_.
https://github.com/microsoft/power-fx#daily-builds

**Reply by cchannon** [reply]
Quick note you might find helpful: When I jumped to this Nuget Stream, the namespace 

```
Microsoft.PowerFx.FeatureFlags
```

(Which I was using for setting StringInterpolation=true following the Samples)

no longer exists. You should probably update the Samples to point us to the right place for this. 

...I don't actually need this for my purposes, just letting you know about it.

**Reply by cchannon** [reply]
Still not the resolution (didn't expect it to be, given @anderson-joyle post, but worth checking)

![image](https://user-images.githubusercontent.com/12833126/206352414-d5c1a50c-6c2a-4427-a5f8-ca861fa30ab8.png)

**Reply by cchannon** [reply]
...but it's nice to be on the latest and greatest! Thank you @MikeStall - I should have caught this way back when I started but I clearly blew past this part of the README!

**Comment by anderson-joyle** [comment, marked as answer]
According to the documentation, there are 3 implementations of Patch function:

**`Patch( DataSource, BaseRecord, ChangeRecord1 [, ChangeRecord2, … ])`** - Implemented
**`Patch( DataSource, BaseRecordsTable, ChangeRecordTable1 [, ChangeRecordTable2, … ] )`** - Not implemented yet
**`Patch( Record1, Record2 [, …] )`** - The one you are trying to execute. Contains known issues.
 
The PR mentioned by @jorisdg is actually a very similar fix to this problem but for a different Patch implementation (the first in the list above).
The following code is similar to yours but execute a different implementation (the first in the list above).

```
var opt = new ParserOptions()
{
    AllowsSideEffects = true,
};

var engine = new RecalcEngine(new PowerFxConfig());

NamedValue[] values = new NamedValue[] {
    new NamedValue(new KeyValuePair<string, FormulaValue>("Id", FormulaValue.New("this is an id"))),
    new NamedValue(new KeyValuePair<string, FormulaValue>("Name", FormulaValue.New("this is a Name"))),
    new NamedValue(new KeyValuePair<string, FormulaValue>("LogicalName", FormulaValue.New("this is a LogicalName")))
};

var thingy = FormulaValue.NewRecordFromFields(values);
var t1 = FormulaValue.NewTable(thingy.Type, thingy); // For this Patch implementation we need a table.

engine.UpdateVariable("t1", t1);
engine.UpdateVariable("thingy", thingy);
engine.Config.SymbolTable.EnableMutationFunctions();

var check = engine.Check("Set(thingy, Patch(t1, thingy, { Id: \"A New and Different Id\"}));thingy.Id", options: opt);

Assert.True(check.IsSuccess);

var result = (StringValue)check.GetEvaluator().Eval();

Assert.Equal("A New and Different Id", result.Value);
```

This problem with `Patch( Record1, Record2 [, …] )` implementation is under our radar. I'll notify you once this has been fixed.

**Reply by cchannon** [reply]
Sorry, no dice. I think I followed pretty closely, but I am getting the same error.

```
NamedValue[] values = new NamedValue[] {
     new NamedValue(new KeyValuePair<string, FormulaValue>("Id", FormulaValue.New("this is an id"))),
     new NamedValue(new KeyValuePair<string, FormulaValue>("Name", FormulaValue.New("this is a Name"))),
     new NamedValue(new KeyValuePair<string, FormulaValue>("LogicalName", FormulaValue.New("this is a LogicalName")))
};

var newRecord = FormulaValue.NewRecordFromFields(values);
var t1 = FormulaValue.NewTable(newRecord.Type, newRecord);
engine.UpdateVariable("t1", t1);
engine.UpdateVariable("myobj", newRecord);
```

![image](https://user-images.githubusercontent.com/12833126/206326710-32ac4983-3e3f-469f-97a6-4727b31f837c.png)

**Reply by cchannon** [reply]
Aha! we have something new now from combining this thread with @MikeStall advice from other comments: 

![image](https://user-images.githubusercontent.com/12833126/206352939-58be1c3d-1f8a-472d-8327-8589dd673b56.png)

I'm not sure what the ![...] and *[...] notations denote, but that seems to be what is tripping me up here

**Reply by anderson-joyle** [reply]
`![...]` represents a record type.
`*[...]` represents a table type.

Your results are interesting. Could you please fork [the repository](https://github.com/microsoft/power-fx-host-samples) and push your changes so I can take a look?

This could be some sort of lagging with the nugget version.

**Reply by cchannon** [reply]
suuuuper cool - just by starting over to give you a clean PR, the exact same Set() operation worked cleanly. I will run a Diff locally and let you know what I find.

**Reply by cchannon** [reply]
OK, I found it - the issue wasn't with PFX: it was in all the Regex matching. I had some test code left in there from when I first started this, trying to isolate any Set() functions so I could declare variables on the fly, but my Regex _wasn't_ actually isolating the Set, it was picking up from the Set to the _next closing paren_. So that regex was quietly chopping off the last closing paren.

Thank you all so much!

**Reply by MikeStall** [reply]
Don't use Regex ... at least, use Parse!

**Reply by cchannon** [reply]
Oh yeah, I completely agree. This was just a holdover fragment from early on in my exploration of the interpreter when I was trying to figure out if I could allow users to dynamically declare variables. I decided it was possible, but not worth the miles of parsing and how impossible all that would be to maintain.
